### PR TITLE
osd: avoid encoding the same log entries repeatedly for different peers

### PR DESCRIPTION
**Pull request by majianpeng**
For replicated pool, the replica size normal is three. Using this PR can reduce once encode(log_entries).

### CODE REVIEW

**Comment by majianpeng** [comment]
retest

**Comment by majianpeng** [comment]
retest this PR

File: src/osd/ReplicatedBackend.cc [hunk 1]
```diff
@@ -961,7 +961,6 @@ MOSDRepOp * ReplicatedBackend::generate_subop(
   eversion_t pg_roll_forward_to,
   hobject_t new_temp_oid,
   hobject_t discard_temp_oid,
-  const vector<pg_log_entry_t> &log_entries,
```
Comment (tchaikov) [inline src/osd/ReplicatedBackend.cc]: could also replace this param with a `buffer::list`, which holds the encoded `log_entries`.

File: src/osd/ReplicatedBackend.cc [hunk 2]
```diff
@@ -1021,45 +1018,51 @@ void ReplicatedBackend::issue_op(
...
-    if (op->op && op->op->pg_trace)
-      wr->trace.init("replicated op", nullptr, &op->op->pg_trace);
-    get_parent()->send_message_osd_cluster(
-      peer.osd, wr, get_osdmap()->get_epoch());
-  }
+      // avoid doing the same work in generate_subop
+      bufferlist logs;
+      encode(log_entries, logs);
+
+      for (set<pg_shard_t>::const_iterator i =
```
Comment (tchaikov) [inline src/osd/ReplicatedBackend.cc]: might want to take this chance to switch to the range-based loop.

**Comment by majianpeng** [comment]
@tchaikov  update by your suggestion. By the way, is it need to encode transaction like log_entries?
For most case, it can reduce once encode(Transaction).

File: src/osd/ReplicatedBackend.cc [hunk 3]
```diff
@@ -1055,7 +1060,7 @@ void ReplicatedBackend::issue_op(
 	    shard,
 	    pinfo);
 
-	if (op->op && op->op->pg_trace)
+	if (pg_trace)
```
Comment (tchaikov) [inline src/osd/ReplicatedBackend.cc]: i don't really like this change, as both op->op and op->op->pg_trace.valid() are O(1), and should complete in a jiffy. introducing another variable does not improve the readability.

**Comment by majianpeng** [comment]
update. please review again. Thanks

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-02-05_11:48:09-rados-wip-kefu-testing-2018-02-05-1650-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2018-02-06_06:20:39-rados-wip-kefu-testing-2018-02-05-1650-distro-basic-mira/

**Pull request merged**